platform-sdk/docs/proposals/001-ledger-state-api/001-ledger-state-api.md

```diff
@@ -0,0 +1,315 @@
+   *
+   * @return The state key. This will never be null, and will always be the same value for an
+   *     instance of {@link ReadableKVState}.
+   */
+  @NonNull
+  String getStateKey();
+
+  /**
+   * Gets the singleton value.
+   *
+   * @return The value, or null if there is no value.
+   */
+  @Nullable
+  T get();
+
+  /**
+   * Gets whether the value of this {@link ReadableSingletonState} has been read.
+   *
+   * @return true if {@link #get()} has been called on this instance
+   */
+  boolean isRead();
+}
+/**
+ * Provides mutable access to singleton state.
+ *
+ * @param <T> The type of the state
+ */
+public interface WritableSingletonState<T> extends ReadableSingletonState<T> {
+  /**
+   * Sets the given value on this state.
+   *
+   * @param value The value. May be null.
+   */
+  void put(@Nullable T value);
+
+  /**
+   * Gets whether the {@link #put(Object)} method has been called on this instance.
+   *
+   * @return True if the {@link #put(Object)} method has been called
+   */
+  boolean isModified();
+}
+```
+
+## Requirements
+
+- `PlatformState` should be accessible via the State API 
+- there should be a new module - `swirlds-state-api` - to host a set of interfaces, records, and abstract classes that represent the Hashgraph state.
+  This module should have a minimal set of dependencies. The Block Node should not have a compile-time dependency
+  on any other modules but this one to interact with the state.
+
+## Changes
+
+### Public API
+
+State API classes and interfaces will migrate to a designated module. The classes themselves will not change. `PlatformState` class will become a part of the public API.
+
+Usage example:
+```java
+private PlatformState findPlatformState(HederaState state) {
+  final ReadableStates states = state.getReadableStates(PlatformState.NAME);   
+  final ReadableSingletonState<PlatformState> platformState = states.getSingleton(PlatformState.PLATFORM_STATE_KEY);
+  return platformState.get();
+}
+```
+
+## Components and Architecture
+
+[### Platform State](#platform-state)
+
+Currently `PlatformState` is a special case node of the merkle tree. That is, it doesn't conform to State API. 
+As a part of the simplification effort **it should be refactored to a singleton state**. 
+As a part of this change, `PlatformState` should not be a `MerkleNode` anymore.
+
+Protobuf definition:
+
+```
+/**
+ * The current state of platform consensus.<br/>
+ * This message stores the current consensus data for the platform
+ * in network state.
+ *
+ * The platform state SHALL represent the latest round's consensus.<br/>
+ * This data SHALL be used to ensure consistency and provide critical data for
+ * restart and reconnect.
+ */
+message PlatformState {
+
+    /**
+     * A version describing the current version of application software.
+     * <p>
+     * This SHALL be the software version that created this state.
+     */
+    proto.SemanticVersion creation_software_version = 1;
```

Comment:
Does this get updated each release? Or only the genesis release? The name `creation_software_version` is not strange to me, I am not sure what it is exactly.

Reply:
Every time a new state object is constructed, this field will be populated with the current software version. Which is to say: yes, this will be updated each release, not just at genesis.

`This SHALL be the software version that created this state.` Do you think that this statement needs to be expanded upon? I can't think of any alternate interpretations that make it ambiguous, but i I'm open to suggestions

Reply:
I think I don't understand the lifecycle of a PlatformState, or what is meant by "every time a new state object is constructed". Does this mean every fast copy? Or does it mean at restart? Or at upgrade? Why is it important to store this version information with the platform state, as the version is available in other ways (and is in another place in state)?

Reply:
> Does this mean every fast copy? Or does it mean at restart? Or at upgrade?

Practically, the value will only be changed at genesis, and at upgrades (since in all other cases, the version of the prior state will just be carried forward.

> and is in another place in state

I wasn't aware that this was stored elsewhere in the state. Bearing that in mind, I can't think of any reason that the duplicate data would need to be in the platform state. Since @imalygin wrote this proposal, I'll hold off making this change until he can weigh in.

Reply:
@alittley i don't think we need to wait for @imalygin to return to decide this. The PlatformState structure was based on the current structure and input from the hashgraph team. If the version is somewhere else in state and is now accessible to the platform, we can use that. I will point out though, that removing it from the platform state will require every single app to include the version in its state. Sure, it's already in the Hedera App state, but is that a requirement we want to add? Maybe the version should stay in the platform and be removed from the app state.

@rbair23 what do you think?

Reply:
Until block streams, we have leniency in modifying this. I'm fine if this goes in. But we should double-check this sometime before it becomes baked in stone forever.

Reply:
I will merge as-is, and we can reconsider this later